src/observers.ts

```diff
@@ -0,0 +1,251 @@
+  TENURE_WEIGHT_TOTAL_BLOCK_COUNT,
+} from './constants';
+import {
+  BlockHeight,
+  DeepReadonly,
+  Gateway,
+  Gateways,
+  RewardDistributions,
+  WeightedObserver,
+} from './types';
+
+export function getEpochBoundariesForHeight({
+  currentBlockHeight,
+  epochBlockLength = new BlockHeight(DEFAULT_EPOCH_BLOCK_LENGTH),
+  epochZeroBlockHeight,
+}: {
+  currentBlockHeight: BlockHeight;
+  epochBlockLength?: BlockHeight;
+  epochZeroBlockHeight: BlockHeight;
+}): {
+  epochStartHeight: BlockHeight;
+  epochEndHeight: BlockHeight;
+} {
+  const epochIndexForCurrentBlockHeight = Math.floor(
+    currentBlockHeight.valueOf() / epochBlockLength.valueOf(),
+  );
+  const epochStartHeight =
+    epochZeroBlockHeight.valueOf() +
+    epochBlockLength.valueOf() * epochIndexForCurrentBlockHeight;
+  return {
+    epochStartHeight: new BlockHeight(epochStartHeight),
+    epochEndHeight: new BlockHeight(
+      epochStartHeight + epochBlockLength.valueOf() - 1,
+    ),
+  };
+}
+
+// TODO: can we confidently us buffers here in non-node environments?
+export async function getEntropyHashForEpoch({
+  epochStartHeight,
+}: {
+  epochStartHeight: BlockHeight;
+}): Promise<Buffer> {
+  // used as we don't have access to Hash object in smartweave executions, so we concat our buffer and hash it at the end
+  let bufferHash: Buffer = Buffer.from('');
+  // We hash multiples block hashes to reduce the chance that someone will
+  // influence the value produced by grinding with excessive hash power.
+  for (let i = 0; i < DEFAULT_NUM_SAMPLED_BLOCKS; i++) {
+    const blockHeight = Math.max(
+      0,
+      epochStartHeight.valueOf() - DEFAULT_SAMPLED_BLOCKS_OFFSET - i,
+    );
+    const path = `/block/height/${blockHeight}`;
+    const data = await SmartWeave.safeArweaveGet(path);
+    const indep_hash = data.indep_hash;
+    // TODO: add regex check on the indep_hash
+    if (!indep_hash) {
+      throw new ContractError(
+        `Block ${blockHeight.valueOf()} has no indep_hash`,
+      );
+    }
+    bufferHash = Buffer.concat([
+      bufferHash,
+      Buffer.from(indep_hash, 'base64url'),
+    ]);
+  }
+  return SmartWeave.arweave.crypto.hash(bufferHash, 'SHA-256');
+}
+
+export function isGatewayLeaving({
+  gateway,
+  currentBlockHeight,
+}: {
+  gateway: DeepReadonly<Gateway>;
+  currentBlockHeight: BlockHeight;
+}): boolean {
+  return (
+    gateway.status === 'leaving' &&
+    gateway.end - GATEWAY_LEAVE_LENGTH <= currentBlockHeight.valueOf()
+  );
+}
+
+export function isGatewayEligibleForDistribution({
+  epochStartHeight,
+  epochEndHeight,
+  gateway,
+}: {
+  epochStartHeight: BlockHeight;
+  epochEndHeight: BlockHeight;
+  gateway: DeepReadonly<Gateway> | undefined;
+}): boolean {
+  // TODO: should a gateway be eligible if it's hidden?
+  if (!gateway) return false;
+  // gateway must have joined before the epoch started, as it affects weighting for distributions
+  const didStartBeforeEpoch = gateway.start < epochStartHeight.valueOf();
```

Comment:
we discussed that this is non-inclusive, as if the gateway starts at the same block height of the epoch, it's tenure weight would be 0 and set it's total weight to 0 as a result. alternatively, if we default it to 1 - then it would have a greater tenure weight then someone who joined before the epoch start. cc @vilenarios

Reply:
per the whitepaper

> To participate in the epoch, a gateway must have already staked IO tokens and joined the network before it starts.